### ISSUE TITLE
FileNotFoundError

### ISSUE THREAD

**Issue by jfallmann** [bioconda/bioconda-recipes#3126]
I know this issue has been addressed recently #3080 , but it's coming up again for me.
Trying to test locally with simulate travis after successful conda build and ./simulate-travis.py --install-requirements results in AttributeError: 'FileNotFoundError' object has no attribute 'read'

BUILD START: perl-text-wrap-2013.0523-0
updating index in: /opt/conda/conda-bld/linux-64
updating index in: /opt/conda/conda-bld/noarch
Traceback (most recent call last):
  File "/opt/conda/bin/conda-build", line 6, in <module>
    sys.exit(conda_build.cli.main_build.main())
  File "/opt/conda/lib/python3.5/site-packages/conda_build/cli/main_build.py", line 242, in main
    execute(sys.argv[1:])
  File "/opt/conda/lib/python3.5/site-packages/conda_build/cli/main_build.py", line 234, in execute
    already_built=None, config=config)
  File "/opt/conda/lib/python3.5/site-packages/conda_build/api.py", line 77, in build
    need_source_download=need_source_download, config=config)
  File "/opt/conda/lib/python3.5/site-packages/conda_build/build.py", line 1099, in build_tree
    config=recipe_config)
  File "/opt/conda/lib/python3.5/site-packages/conda_build/build.py", line 634, in build
    create_env(config.build_prefix, specs, config=config)
  File "/opt/conda/lib/python3.5/site-packages/conda_build/build.py", line 494, in create_env
    index = get_build_index(config=config, clear_cache=True)
  File "/opt/conda/lib/python3.5/site-packages/conda_build/build.py", line 449, in get_build_index
    use_cache=not clear_cache)
  File "/opt/conda/lib/python3.5/site-packages/conda/api.py", line 24, in get_index
    index = fetch_index(channel_urls, use_cache=use_cache, unknown=unknown)
  File "/opt/conda/lib/python3.5/site-packages/conda/fetch.py", line 300, in fetch_index
    repodatas = [(u, f.result()) for u, f in zip(urls, futures)]
  File "/opt/conda/lib/python3.5/site-packages/conda/fetch.py", line 300, in <listcomp>
    repodatas = [(u, f.result()) for u, f in zip(urls, futures)]
  File "/opt/conda/lib/python3.5/concurrent/futures/_base.py", line 398, in result
    return self.__get_result()
  File "/opt/conda/lib/python3.5/concurrent/futures/_base.py", line 357, in __get_result
    raise self._exception
  File "/opt/conda/lib/python3.5/concurrent/futures/thread.py", line 55, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/opt/conda/lib/python3.5/site-packages/conda/fetch.py", line 75, in func
    res = f(*args, **kwargs)
  File "/opt/conda/lib/python3.5/site-packages/conda/fetch.py", line 117, in fetch_repodata
    timeout=(6.1, 60))
  File "/opt/conda/lib/python3.5/site-packages/requests/sessions.py", line 501, in get
    return self.request('GET', url, **kwargs)
  File "/opt/conda/lib/python3.5/site-packages/requests/sessions.py", line 488, in request
    resp = self.send(prep, **send_kwargs)
  File "/opt/conda/lib/python3.5/site-packages/requests/sessions.py", line 641, in send
    r.content
  File "/opt/conda/lib/python3.5/site-packages/requests/models.py", line 772, in content
    self._content = bytes().join(self.iter_content(CONTENT_CHUNK_SIZE)) or bytes()
  File "/opt/conda/lib/python3.5/site-packages/requests/models.py", line 705, in generate
    chunk = self.raw.read(chunk_size)
AttributeError: 'FileNotFoundError' object has no attribute 'read'

ERROR bioconda_utils.build:build_recipes(355): BIOCONDA BUILD SUMMARY: of 1 recipes, 1 failed and 0 were skipped. Details of recipes and environments follow.
ERROR bioconda_utils.build:build_recipes(366): BIOCONDA BUILD SUMMARY: FAILED recipe perl-text-wrap-2013.0523-0.tar.bz2, environment CONDA_GSL=1.16;CONDA_BOOST=1.60;CONDA_NPY=110;CONDA_GMP=5.1;CONDA_R=3.3.1;CONDA_PY=27;CONDA_NCURSES=5.9;CONDA_PERL=5.22.0
Traceback (most recent call last):
  File "./simulate-travis.py", line 150, in <module>
    sp.run(['scripts/travis-run.sh'], env=env, universal_newlines=True, check=True)
  File "/homes/biertruck/fall/anaconda3/lib/python3.5/subprocess.py", line 708, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['scripts/travis-run.sh']' returned non-zero exit status 1

I'm pretty sure the error was gone already after #3080 and now it's back, any idea what happened
or is it something I'm doing wrong?

**Comment by daler** [comment]
Can you try rebasing to master, or from a fresh clone from https://github.com/bioconda/bioconda-recipes.git?

**Comment by jfallmann** [comment]
@daler Fresh clone gets rid of the error, not sure why this is not working when fetching from upstream and merging into my fork, however, problem solved, new problem, but this seems to be some docker related stuff, thanks for your help again!

ERROR bioconda_utils.docker_utils:_build_image(350): BIOCONDA DOCKER FAIL: Error building docker container tmp-bioconda-builder. stdout+stdterr follows:
ERROR bioconda_utils.docker_utils:_build_image(351): None
Traceback (most recent call last):
  File "/home/fall/anaconda3/bin/bioconda-utils", line 11, in <module>
    load_entry_point('bioconda-utils==0.9.0', 'console_scripts', 'bioconda-utils')()
  File "/home/fall/anaconda3/lib/python3.5/site-packages/bioconda_utils/cli.py", line 172, in main
    argh.dispatch_commands([build, dag, dependent])
  File "/home/fall/anaconda3/lib/python3.5/site-packages/argh/dispatching.py", line 328, in dispatch_commands
    dispatch(parser, *args, **kwargs)
  File "/home/fall/anaconda3/lib/python3.5/site-packages/argh/dispatching.py", line 174, in dispatch
    for line in lines:
  File "/home/fall/anaconda3/lib/python3.5/site-packages/argh/dispatching.py", line 277, in _execute_command
    for line in result:
  File "/home/fall/anaconda3/lib/python3.5/site-packages/argh/dispatching.py", line 260, in _call
    result = function(*positional, **keywords)
  File "/home/fall/anaconda3/lib/python3.5/site-packages/bioconda_utils/cli.py", line 92, in build
    conda_build_version=conda_build_version,
  File "/home/fall/anaconda3/lib/python3.5/site-packages/bioconda_utils/docker_utils.py", line 302, in __init__
    self._build_image()
  File "/home/fall/anaconda3/lib/python3.5/site-packages/bioconda_utils/docker_utils.py", line 352, in _build_image
    raise e
  File "/home/fall/anaconda3/lib/python3.5/site-packages/bioconda_utils/docker_utils.py", line 346, in _build_image
    p = utils.run(cmd)
  File "/home/fall/anaconda3/lib/python3.5/site-packages/bioconda_utils/utils.py", line 68, in run
    raise e
  File "/home/fall/anaconda3/lib/python3.5/site-packages/bioconda_utils/utils.py", line 64, in run
    p = sp.run(cmds, stdout=sp.PIPE, stderr=sp.STDOUT, check=True, env=env)
  File "/home/fall/anaconda3/lib/python3.5/subprocess.py", line 708, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['docker', 'build', '-t', 'tmp-bioconda-builder', '/tmp/tmpt2z2wdun']' returned non-zero exit status 1
Traceback (most recent call last):
  File "./simulate-travis.py", line 150, in <module>
    sp.run(['scripts/travis-run.sh'], env=env, universal_newlines=True, check=True)
  File "/home/fall/anaconda3/lib/python3.5/subprocess.py", line 708, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['scripts/travis-run.sh']' returned non-zero exit status 1

**Comment by bgruening** [comment]
@jfallmann is Docker running on your host without sudo?

**Comment by jfallmann** [comment]
@bgruening Yeah it does, created the not recommended docker group and added myself to it, 
but it doesn't really matter, same error with sudo, other docker containers run fine, e.g. the rna-workbench container.
Running bioconda-utils build recipes config.yml --docker from terminal results in the same error,
no matter if sudo or not. Any Idea where I can start looking for error messages that could help resolve the problems?

**Comment by bgruening** [comment]
Not really, it's dying at this step: 'docker build -t tmp-bioconda-builder /tmp/tmpt2z2wdun'.
U haven't seen this before, you have enough space? Can you build other containers?

**Comment by jfallmann** [comment]
Well it's the first time i've come this far :( 
Space is 8GB on /tmp and >100GB in home so this is hopefully not the problem,
Successfully built b2c5cde9a844
Totally lost atm, Travis at least builds and then starts complaining ;)
If you have any idea let me know
Thx!!!

**Comment by daler** [comment]
@jfallmann can you run `simulate-travis.py` with the `--loglevel=debug` option and make the log available somewhere? (it might be big)

**Comment by jfallmann** [comment]
@daler Unfortunately it's not big at all, trying this at home and on my workstation machine, so far similar problems

./simulate-travis.py --packages perl-text-wrap --loglevel=debug
+ bioconda-utils build recipes config.yml --docker --loglevel=info --mulled-test --packages perl-text-wrap --loglevel=debug
INFO bioconda_utils.docker_utils:_pull_image(308): BIOCONDA DOCKER: Pulling docker image condaforge/linux-anvil
DEBUG bioconda_utils.docker_utils:_pull_image(310): BIOCONDA DOCKER: stdout+stderr: Using default tag: latest
Trying to pull repository docker.io/condaforge/linux-anvil ... 
latest: Pulling from docker.io/condaforge/linux-anvil
Digest: sha256:cb49695706568453d2b76eb6160eaec9b65832c49a261b8a0d6c6a800c4e7692
Status: Image is up to date for docker.io/condaforge/linux-anvil:latest

INFO bioconda_utils.docker_utils:_pull_image(314): BIOCONDA DOCKER: Done pulling image
DEBUG bioconda_utils.docker_utils:_build_image(325): BIOCONDA DOCKER: Building image "tmp-bioconda-builder" from /tmp/tmp7gi3i2g2
DEBUG bioconda_utils.docker_utils:_build_image(338): Dockerfile:

FROM condaforge/linux-anvil
COPY requirements.txt /tmp/requirements.txt
RUN /opt/conda/bin/conda install --file /tmp/requirements.txt

ERROR bioconda_utils.docker_utils:_build_image(350): BIOCONDA DOCKER FAIL: Error building docker container tmp-bioconda-builder. stdout+stdterr follows:
ERROR bioconda_utils.docker_utils:_build_image(351): None
Traceback (most recent call last):
  File "/home/fall/anaconda3/bin/bioconda-utils", line 11, in <module>
    load_entry_point('bioconda-utils==0.9.0', 'console_scripts', 'bioconda-utils')()
  File "/home/fall/anaconda3/lib/python3.5/site-packages/bioconda_utils/cli.py", line 172, in main
    argh.dispatch_commands([build, dag, dependent])
  File "/home/fall/anaconda3/lib/python3.5/site-packages/argh/dispatching.py", line 328, in dispatch_commands
    dispatch(parser, *args, **kwargs)
  File "/home/fall/anaconda3/lib/python3.5/site-packages/argh/dispatching.py", line 174, in dispatch
    for line in lines:
  File "/home/fall/anaconda3/lib/python3.5/site-packages/argh/dispatching.py", line 277, in _execute_command
    for line in result:
  File "/home/fall/anaconda3/lib/python3.5/site-packages/argh/dispatching.py", line 260, in _call
    result = function(*positional, **keywords)
  File "/home/fall/anaconda3/lib/python3.5/site-packages/bioconda_utils/cli.py", line 92, in build
    conda_build_version=conda_build_version,
  File "/home/fall/anaconda3/lib/python3.5/site-packages/bioconda_utils/docker_utils.py", line 302, in __init__
    self._build_image()
  File "/home/fall/anaconda3/lib/python3.5/site-packages/bioconda_utils/docker_utils.py", line 352, in _build_image
    raise e
  File "/home/fall/anaconda3/lib/python3.5/site-packages/bioconda_utils/docker_utils.py", line 346, in _build_image
    p = utils.run(cmd)
  File "/home/fall/anaconda3/lib/python3.5/site-packages/bioconda_utils/utils.py", line 68, in run
    raise e
  File "/home/fall/anaconda3/lib/python3.5/site-packages/bioconda_utils/utils.py", line 64, in run
    p = sp.run(cmds, stdout=sp.PIPE, stderr=sp.STDOUT, check=True, env=env)
  File "/home/fall/anaconda3/lib/python3.5/subprocess.py", line 708, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['docker', 'build', '-t', 'tmp-bioconda-builder', '/tmp/tmp7gi3i2g2']' returned non-zero exit status 1
Traceback (most recent call last):
  File "./simulate-travis.py", line 150, in <module>
    sp.run(['scripts/travis-run.sh'], env=env, universal_newlines=True, check=True)
  File "/home/fall/anaconda3/lib/python3.5/subprocess.py", line 708, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['scripts/travis-run.sh']' returned non-zero exit status 1

**Comment by jfallmann** [comment]
@daler Oh no, I could have tried this already, seems there is the error, 
PackageNotFoundError: Package not found: '' Packages missing in current linux-64 channels: 
  - galaxy-lib 16.10.9 py35_0
  - involucro 1.1.2 0

But this can't be a problem that's coming from my machines, these should be part part of the docker image or am I wrong? Anyway, thanks a lot for your time!!!

Here the whole output
docker build -t tmp-bioconda-builder /tmp/tmp7gi3i2g2
Sending build context to Docker daemon 3.072 kB
Step 1 : FROM condaforge/linux-anvil
 ---> a4ca686f3748
Step 2 : COPY requirements.txt /tmp/requirements.txt
 ---> Using cache
 ---> ccb0e0312cd3
Step 3 : RUN /opt/conda/bin/conda install --file /tmp/requirements.txt
 ---> Running in c682df966f7b
Fetching package metadata .........
.Solving package specifications: .


PackageNotFoundError: Package not found: '' Packages missing in current linux-64 channels: 
  - galaxy-lib 16.10.9 py35_0
  - involucro 1.1.2 0

You can search for packages on anaconda.org with

    anaconda search -t conda involucro

(and similarly for the other packages)

You may need to install the anaconda-client command line client with

    conda install anaconda-client


The command '/bin/sh -c /opt/conda/bin/conda install --file /tmp/requirements.txt' returned a non-zero code: 1

**Comment by daler** [comment]
@jfallmann can you try these commands in order:

```
./simulate-travis.py --set-channel-order
./simulate-travis.py --install-requirements
./simulate-travis.py --quick --loglevel=debug
```

**Comment by jfallmann** [comment]
@daler Same result, channel order and requirements were fine, as this happens at home and at work maybe it's something specific for my setup, they are as similar as I could get them,
I'll see if I can get a clean environment for testing today to be sure this isn't a problem on my side

$ ./simulate-travis.py --set-channel-order

          Warnings like "'conda-forge' already in 'channels' list, moving to the top"
          are expected if channels have been added before, and can be safely ignored.
          
Warning: 'conda-forge' already in 'channels' list, moving to the top
Warning: 'defaults' already in 'channels' list, moving to the top
Warning: 'r' already in 'channels' list, moving to the top
Warning: 'bioconda' already in 'channels' list, moving to the top

conda config is now:

--add channels 'conda-forge'   # lowest priority
--add channels 'defaults'
--add channels 'r'
--add channels 'bioconda'   # highest priority

##############

./simulate-travis.py --install-requirements
Fetching package metadata .............
.Solving package specifications: ..........

# All requested packages already installed.
# packages in environment at /home/fall/anaconda3:
#
anaconda-client           1.5.2                    py35_0  
argh                      0.26.2                   py35_0    bioconda
beautifulsoup4            4.4.1                    py35_0  
conda                     4.2.9                    py35_0  
conda-build               2.0.7                    py35_0  
galaxy-lib                16.10.9                  py35_0    bioconda
involucro                 1.1.2                         0    bioconda
jsonschema                2.5.1                    py35_0  
networkx                  1.11                     py35_0  
pyaml                     15.8.2                   py35_0    bioconda
pydotplus                 2.0.2                    py35_0    bioconda
python                    3.5.2                         0  
requests                  2.10.0                   py35_0  
sphinx                    1.4.1                    py35_0  
Collecting git+https://github.com/bioconda/bioconda-utils.git@reqs-updates
  Cloning https://github.com/bioconda/bioconda-utils.git (to reqs-updates) to /tmp/pip-86ptd4n0-build
  Requirement already satisfied (use --upgrade to upgrade): bioconda-utils==0.9.0 from git+https://github.com/bioconda/bioconda-utils.git@reqs-updates in /home/fall/anaconda3/lib/python3.5/site-packages
Requirement already satisfied: argh in /home/fall/anaconda3/lib/python3.5/site-packages (from bioconda-utils==0.9.0)
Requirement already satisfied: networkx in /home/fall/anaconda3/lib/python3.5/site-packages (from bioconda-utils==0.9.0)
Requirement already satisfied: pydotplus in /home/fall/anaconda3/lib/python3.5/site-packages (from bioconda-utils==0.9.0)
Requirement already satisfied: pyyaml in /home/fall/anaconda3/lib/python3.5/site-packages (from bioconda-utils==0.9.0)
Requirement already satisfied: requests<2.11 in /home/fall/anaconda3/lib/python3.5/site-packages (from bioconda-utils==0.9.0)
Requirement already satisfied: jsonschema in /home/fall/anaconda3/lib/python3.5/site-packages (from bioconda-utils==0.9.0)
Requirement already satisfied: decorator>=3.4.0 in /home/fall/anaconda3/lib/python3.5/site-packages (from networkx->bioconda-utils==0.9.0)
Requirement already satisfied: pyparsing>=2.0.1 in /home/fall/anaconda3/lib/python3.5/site-packages (from pydotplus->bioconda-utils==0.9.0)

###############

./simulate-travis.py --quick --loglevel=debug
+ bioconda-utils build recipes config.yml --docker --loglevel=info --mulled-test --quick --loglevel=debug
INFO bioconda_utils.docker_utils:_pull_image(308): BIOCONDA DOCKER: Pulling docker image condaforge/linux-anvil
DEBUG bioconda_utils.docker_utils:_pull_image(310): BIOCONDA DOCKER: stdout+stderr: Using default tag: latest
Trying to pull repository docker.io/condaforge/linux-anvil ... 
latest: Pulling from docker.io/condaforge/linux-anvil
Digest: sha256:cb49695706568453d2b76eb6160eaec9b65832c49a261b8a0d6c6a800c4e7692
Status: Image is up to date for docker.io/condaforge/linux-anvil:latest

INFO bioconda_utils.docker_utils:_pull_image(314): BIOCONDA DOCKER: Done pulling image
DEBUG bioconda_utils.docker_utils:_build_image(325): BIOCONDA DOCKER: Building image "tmp-bioconda-builder" from /tmp/tmp6fs9_n32
DEBUG bioconda_utils.docker_utils:_build_image(338): Dockerfile:

FROM condaforge/linux-anvil
COPY requirements.txt /tmp/requirements.txt
RUN /opt/conda/bin/conda install --file /tmp/requirements.txt

ERROR bioconda_utils.docker_utils:_build_image(350): BIOCONDA DOCKER FAIL: Error building docker container tmp-bioconda-builder. stdout+stdterr follows:
ERROR bioconda_utils.docker_utils:_build_image(351): None
Traceback (most recent call last):
  File "/home/fall/anaconda3/bin/bioconda-utils", line 11, in <module>
    load_entry_point('bioconda-utils==0.9.0', 'console_scripts', 'bioconda-utils')()
  File "/home/fall/anaconda3/lib/python3.5/site-packages/bioconda_utils/cli.py", line 172, in main
    argh.dispatch_commands([build, dag, dependent])
  File "/home/fall/anaconda3/lib/python3.5/site-packages/argh/dispatching.py", line 328, in dispatch_commands
    dispatch(parser, *args, **kwargs)
  File "/home/fall/anaconda3/lib/python3.5/site-packages/argh/dispatching.py", line 174, in dispatch
    for line in lines:
  File "/home/fall/anaconda3/lib/python3.5/site-packages/argh/dispatching.py", line 277, in _execute_command
    for line in result:
  File "/home/fall/anaconda3/lib/python3.5/site-packages/argh/dispatching.py", line 260, in _call
    result = function(*positional, **keywords)
  File "/home/fall/anaconda3/lib/python3.5/site-packages/bioconda_utils/cli.py", line 92, in build
    conda_build_version=conda_build_version,
  File "/home/fall/anaconda3/lib/python3.5/site-packages/bioconda_utils/docker_utils.py", line 302, in __init__
    self._build_image()
  File "/home/fall/anaconda3/lib/python3.5/site-packages/bioconda_utils/docker_utils.py", line 352, in _build_image
    raise e
  File "/home/fall/anaconda3/lib/python3.5/site-packages/bioconda_utils/docker_utils.py", line 346, in _build_image
    p = utils.run(cmd)
  File "/home/fall/anaconda3/lib/python3.5/site-packages/bioconda_utils/utils.py", line 68, in run
    raise e
  File "/home/fall/anaconda3/lib/python3.5/site-packages/bioconda_utils/utils.py", line 64, in run
    p = sp.run(cmds, stdout=sp.PIPE, stderr=sp.STDOUT, check=True, env=env)
  File "/home/fall/anaconda3/lib/python3.5/subprocess.py", line 708, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['docker', 'build', '-t', 'tmp-bioconda-builder', '/tmp/tmp6fs9_n32']' returned non-zero exit status 1
Traceback (most recent call last):
  File "./simulate-travis.py", line 150, in <module>
    sp.run(['scripts/travis-run.sh'], env=env, universal_newlines=True, check=True)
  File "/home/fall/anaconda3/lib/python3.5/subprocess.py", line 708, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['scripts/travis-run.sh']' returned non-zero exit status 1

**Comment by jfallmann** [comment]
@daler Did a fresh install of everything, anaconda, docker, bioconda, reran 
./simulate-travis.py --set-channel-order
./simulate-travis.py --install-requirements
./simulate-travis.py --quick --loglevel=debug

Same error, also running docker build -t tmp-bioconda-builder /tmp/tmp3_yupkoe by hand
results in PackageNotFoundError: Package not found: '' Packages missing in current linux-64 channels: 
  - galaxy-lib 16.10.9 py35_0
  - involucro 1.1.2 0

You can search for packages on anaconda.org with

    anaconda search -t conda involucro

(and similarly for the other packages)

You may need to install the anaconda-client command line client with

    conda install anaconda-client


So I guess it's not a problem on my side

**Comment by daler** [comment]
@jfallmann maybe you have an issue in your local path, or maybe something in `.condarc`? Or maybe some local proxy issues?

I just did the following on a fresh Ubuntu 16.04 AWS instance:

```bash

# get bioconda-recipes and install conda
git clone https://github.com/bioconda/bioconda-recipes.git
wget https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh
bash Miniconda3-latest-Linux-x86_64.sh -b
echo "PATH=\$HOME/miniconda3/bin:\$PATH" >> ~/.bashrc
source ~/.bashrc

#  docker
sudo wget -qO- https://get.docker.com | sh

# was getting error "sudo: unable to resolve host ip-xxx-xx-x-xxx" 
#  (see http://stackoverflow.com/questions/33441873/aws-error-sudo-unable-to-resolve-host-ip-10-0-xx-xx)
echo "127.0.0.1 ip-xxx-xx-x-xxx" | sudo tee -a /etc/hosts

# try again
sudo wget -qO- https://get.docker.com | sh

# let ubuntu user run docker
sudo usermod -aG docker ubuntu

# exit ssh, reconnect, then:

cd bioconda-recipes
./simulate-travis.py --set-channel-order
./simulate-travis.py --install-requirements
./simulate-travis.py --quick --loglevel=debug
```

and it successfully completed, ending with the following log output:

```
DEBUG bioconda_utils.utils:newly_unblacklisted(404): Examining master branch config file
DEBUG bioconda_utils.utils:newly_unblacklisted(417): Recipes newly unblacklisted:

DEBUG bioconda_utils.build:build_recipes(239): Unblacklisted: []
DEBUG bioconda_utils.build:build_recipes(244): Changed: ['recipes/pgma-simple', 'recipes/pgma-simple']
DEBUG bioconda_utils.build:build_recipes(246): Quick-filtered recipes: set()
INFO bioconda_utils.build:build_recipes(248): Filtering recipes
DEBUG bioconda_utils.utils:filter_recipes(494): recipes: set()
INFO bioconda_utils.build:build_recipes(262): Nothing to be done.

```

**Comment by jfallmann** [comment]
@daler @bgruening 
Sanitized bash (unset all env vars, aliases, cleaned path), .condarc only contains 
channels:
  - bioconda
  - r
  - defaults
  - conda-forge

still the same problem.
As I do not know of any proxy settings that would concern my home and my work setup, I don't think that's the problem either.
However, as this is eating a lot of time better used to actually contribute to the repo, I'll postpone this for now and rely on travis.
Thanks a lot for your time and help!

**Comment by daler** [comment]
@jfallmann sorry we're not able to solve it. You could always spin up a VM or use AWS, or worst-case, wait for travis-ci to run the tests. If the latter, it will help keep resources free if you do that kind of testing on your own fork of the repo.

And if you figure out what the issue is please let us know in case others have similar issues.

**Comment by jfallmann** [comment]
@daler no worries, I'll probably just go with a VM, if I find out what's causing this problem I'll let you know, and thanks for the tip with the travis test on my own fork, makes sense, but whenever I push to my fork now travis automatically runs a test on the main repo, i guess because of the open pull request?

**Comment by bgruening** [comment]
@jfallmann the open PR will trigger also a test in bioconda, yes.

**Comment by jfallmann** [comment]
@bgruening I guess then it's better to close the PR and open a new one once Travis builds on my branch

**Comment by bgruening** [comment]
Sounds good. I'm still not sure what the problem here is and others seems to have it as well :(

**Comment by jfallmann** [comment]
The PackageNotFoundError is quite confusing, I currently have no idea where to start debugging on my side, but this just means I have more time for other stuff during Travis builds ;)

**Comment by tetianakh** [comment]
Here's what I've figured out:
The conda inside docker container we're building doesn't know about additional channels.
I've changed the dockerfile template in bioconda_utils/docker_utils.py (the  bioconda-utils package) to:

```FROM condaforge/linux-anvil
COPY requirements.txt /tmp/requirements.txt
RUN /opt/conda/bin/conda config --add channels conda-forge
RUN /opt/conda/bin/conda config --add channels bioconda
RUN /opt/conda/bin/conda config --add channels r
RUN /opt/conda/bin/conda config --add channels defaults
RUN /opt/conda/bin/conda --version
RUN /opt/conda/bin/conda install --file /tmp/requirements.txt
```

And now everything works. 

Maybe someone with a better knowledge can understand this better.

**Comment by jfallmann** [comment]
This is nice!!! And it explains the error I get,  is this already available?

**Comment by jfallmann** [comment]
I commited your changes to my fork of bioconda-utils and reinstalled from that fork and I got rid of this FileNotFound error!
Problem is I now have a new one, complaining about "The remote server could not find the channel you requested" later on. But thats a new problem, thanks for solving the last one!

Traceback (most recent call last):
  File "/opt/conda/lib/python3.5/site-packages/conda/fetch.py", line 120, in fetch_repodata
    resp.raise_for_status()
  File "/opt/conda/lib/python3.5/site-packages/requests/models.py", line 844, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 404 Client Error: None for url: file:///opt/host-conda-bld/linux-64/repodata.json.bz2

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/conda/bin/conda-build", line 6, in <module>
    sys.exit(conda_build.cli.main_build.main())
  File "/opt/conda/lib/python3.5/site-packages/conda_build/cli/main_build.py", line 242, in main
    execute(sys.argv[1:])
  File "/opt/conda/lib/python3.5/site-packages/conda_build/cli/main_build.py", line 234, in execute
    already_built=None, config=config)
  File "/opt/conda/lib/python3.5/site-packages/conda_build/api.py", line 77, in build
    need_source_download=need_source_download, config=config)
  File "/opt/conda/lib/python3.5/site-packages/conda_build/build.py", line 1099, in build_tree
    config=recipe_config)
  File "/opt/conda/lib/python3.5/site-packages/conda_build/build.py", line 634, in build
    create_env(config.build_prefix, specs, config=config)
  File "/opt/conda/lib/python3.5/site-packages/conda_build/build.py", line 494, in create_env
    index = get_build_index(config=config, clear_cache=True)
  File "/opt/conda/lib/python3.5/site-packages/conda_build/build.py", line 449, in get_build_index
    use_cache=not clear_cache)
  File "/opt/conda/lib/python3.5/site-packages/conda/api.py", line 24, in get_index
    index = fetch_index(channel_urls, use_cache=use_cache, unknown=unknown)
  File "/opt/conda/lib/python3.5/site-packages/conda/fetch.py", line 300, in fetch_index
    repodatas = [(u, f.result()) for u, f in zip(urls, futures)]
  File "/opt/conda/lib/python3.5/site-packages/conda/fetch.py", line 300, in <listcomp>
    repodatas = [(u, f.result()) for u, f in zip(urls, futures)]
  File "/opt/conda/lib/python3.5/concurrent/futures/_base.py", line 398, in result
    return self.__get_result()
  File "/opt/conda/lib/python3.5/concurrent/futures/_base.py", line 357, in __get_result
    raise self._exception
  File "/opt/conda/lib/python3.5/concurrent/futures/thread.py", line 55, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/opt/conda/lib/python3.5/site-packages/conda/fetch.py", line 75, in func
    res = f(*args, **kwargs)
  File "/opt/conda/lib/python3.5/site-packages/conda/fetch.py", line 223, in fetch_repodata
    e.response.reason if e.response else None)
conda.exceptions.CondaHTTPError: HTTP 404 None
for url <None>

The remote server could not find the channel you requested.

You will need to adjust your conda configuration to proceed.
Use `conda config --show` to view your configuration's current state.
Further configuration help can be found at <http://conda.pydata.org/docs/config.html>.

**Comment by daler** [comment]
@jfallmann @tetianakh 

Running `simulate-travis.py --install-requirements` does this:
- finds the configured branch using the `BIOCONDA_UTILS_TAG` variable in `.travis.yml`. Currently bioconda-recipes master branch is set to use the `reqs-update` branch (see [here](https://github.com/bioconda/bioconda-recipes/blob/master/.travis.yml#L14)).
- installs the requirements for that branch as well as the branch itself ([here](https://github.com/bioconda/bioconda-recipes/blob/master/simulate-travis.py#L114) is where that happens in simulate-travis.py).

On `reqs-update` branch, the edits you mention are already in the Dockerfile (see [here](https://github.com/bioconda/bioconda-utils/blob/reqs-updates/bioconda_utils/docker_utils.py#L132), note you'll need to match the order of channels).

So it seems like the issue is `simulate-travis.py --install-requirements` is not installing the correct branch of bioconda-utils. I don't immediately see why that's the case but will take a closer look when I get a chance.

**Comment by daler** [comment]
@jfallmann try, outside of simulate-travis.py, doing a `conda-build` on a recipe. The `argh` recipe is pretty small, you can try that one.

The reason is that if conda-build was never run on the host, the build directory won't exist. The docker mount then becomes an empty directory and conda doesn't like that as a channel. bioconda-utils should do some checks for the presence of files (repodata.json, probably) and if it's not found maybe build a dummy recipe or something on the host before running in the docker container.

**Comment by tetianakh** [comment]
@daler as far as I can tell, `simulate-travis.py --install-requirements` installs requirements in the host conda, not inside the container. Is that so?

@jfallmann I don't think that hardcoding a list of channels into a Dockerfile template is a good way to resolve this problem. The channels must come from the config.yaml (in case `simulate-travis.py` is run), or from the command line arguments (in case `bioconda-utils build` command is run directly). That's why I've posted the info about Dockerfile hack here, and didn't create a PR to bioconda-utils repo.

Another important question is, why does it work in some cases?

**Comment by jfallmann** [comment]
@daler build before simulate travis did not resolve the issue

@tetianakh I don't think its a good idea either, that's why i tried this on my fork, will not create a PR for that hack, but it would have been nice to have a workaround until this is resolved

I'm also surprised this works for some people and not for others, if it's a problem with the container it shouldn't work at all?

**Comment by daler** [comment]
@tetianakh right, `--install-requirements` only installs bioconda-utils on the host. That's intentional. The only thing we want in the container is conda itself and the channel requirements. 

Currently the Dockerfile in the `reqs-update` branch *does* hard-code the channels.

Ideally channels should be stored in a single place (yaml in bioconda-utils most likely) and that list of channels should be unified across bioconda-utils and bioconda-recipes in both bash travis-setup scripts (which use `conda config --add channel` and need low-to-high priority), command-line calls (high-to-low order) in python modules, and dockerfiles. We haven't hit on a good solution for that.

I agree with you both that there's something we're missing causing it to fail in some cases but not others.

**Comment by tetianakh** [comment]
@jfallmann I didn't have any problems after adding the channels into Dockerfile template on my local setup. But then I've decided to make a clean Ubuntu VM to see if I can reproduce the channels problem. I've done bascally the same things that @daler in a comment where he described that everything works on a fresh Ubuntu EC2 instance. And, I've got the  `CondaHTTPError: HTTP 404 None` error.

After poking around for some time, I've figured out why it happens. `bioconda_utils` builds a docker image and runs a container from it with the following command:
```
docker run --net host --rm \
-v /tmp/tmpkvblcdvo/build_script.bash:/opt/build_script.bash \
-v /home/quitka/miniconda3/conda-bld:/opt/host-conda-bld \
-v /home/quitka/biopconda-recipes/recipes/ericscript:/opt/recipe \
-e CONDA_GSL=1.16 \
-e CONDA_BOOST=1.60 \
-e CONDA_R=3.3.1 \
-e CONDA_PY=27 \
-e CONDA_NPY=110 \
-e CONDA_NCURSES=5.9 \
-e CONDA_GMP=5.1 \
-e CONDA_PERL=5.22.0 \
tmp-bioconda-builder \
/bin/bash /opt/build_script.bash'
```
The important part is that `bioconda_utils` runs a `build_script.bash` file inside the docker container. In that file, there's a line `conda config --add channels file:///opt/host-conda-bld`. And `/opt/host-conda-bld` is a directory to which the host's `conda-bld` directory is mounted. 

Bear with me, we're close to the root of the problem.

It turns out that `conda` doesn't support `file://` URLs in a consistent manner. If the `conda-bld` directory does not exist, **or if it exists but contains no builds**,  `conda` won't understand the `file:// URL`. I can reproduce this in a simple way on any of my `conda` installations:

```
(root) ~$ rm -rf /home/quitka/miniconda3/conda-bld/
(root) ~$ conda config --add channels /home/quitka/miniconda3/conda-bld/
(root) ~$ conda update --all
Fetching package metadata ..............

CondaHTTPError: HTTP 404 None
for url <None>

The remote server could not find the channel you requested.

You will need to adjust your conda configuration to proceed.
Use `conda config --show` to view your configuration's current state.
Further configuration help can be found at <http://conda.pydata.org/docs/config.html>.
```

There's a closed issue in the conda repo: https://github.com/conda/conda/issues/2845
They recommend to "build a package first" to avoid this problem. However, we'd have to build the package inside the container first, then add the `file://...` channel, and then build the package that is being tested -- which is super awkward and wasteful.

For me, everything worked fine after removing the "offending" line from the `BUILD_SCRIPT_TEMPLATE` in `bioconda_utils/docker_utils.py`.

**Does anyone know why we need to add the host's `conda-bld` to channels?**

**Comment by jfallmann** [comment]
Hi @tetianakh, I've read about this problem in  https://github.com/conda/conda/issues/2845 but it had no effect on my problem building a package first, however, as you said it would have to happen inside the directory thats mounted later, which I never did.
Debugging this kind of error is not straight forward, so thanks for finding out the problem and your fix, I'll try and tell you if this works for me too.

**Comment by tetianakh** [comment]
@jfallmann I've written a recipe and the CI build for it fails... I'm trying to reproduce the error locally, and encounter all kinds of issues except the one I'm looking for. So I'm only glad that some of the debugging can be useful for others :smile:

**Comment by tetianakh** [comment]
By the way, here's a little ansible playbook I've made to set up a fresh Ubuntu 16.04 VM: https://gist.github.com/tetianakh/dd619db896fc148d7a9767c2c0af1d10 

Maybe someone will find it useful

**Comment by jfallmann** [comment]
@tetianakh Your fix worked for me, at least for the error I had, got a new one during build 

Fixing permissions
File lock on /opt/recipe/.conda_lock could not be obtained.  You might need to try fewer builds at once.  Otherwise, run conda clean --lock

It's something I've seen somewhere already, will keep digging and come back when this is resolved

**Comment by daler** [comment]
@tetianakh I feel bad you're having to reverse-engineer, I haven't had time for docs on this. But you're exactly right.

> Does anyone know why we need to add the host's conda-bld directory to channels? Can we avoid this step for the sake of reproducibility?

The reason we need the host's conda-build directory is that many times we have multiple interdependent recipes in a single PR. At the end of each successful build by the container, the built package is copied over to the host's build dir. The next container to run mounts that directory and uses it as a channel so that it can use packages previously built in the same PR as dependencies. If we didn't do this, we would need to ensure PRs contain independent recipes. We made the choice to maximize our usage of allocated travis-ci resources.

Can you fool conda and avoid the 404 if you do a `touch $conda-bld-dir/linux-64/repodata{.json,.json.bz2}` in the buildscript.bash?

Also, now that you're digging into the bioconda-utils code, any suggestions/fixes/improvements would be very welcome!

**Comment by tetianakh** [comment]
@daler `touch`ing `$conda-bld-dir/linux-64/repodata{.json,.json.bz2}` doesn't work:

```
Runtime error: Invalid index file: file:///opt/host-conda-bld/linux-64/repodata.json.bz2: Expecting value: line 1 column 1
```

Maybe we can build a minimal package on the host? Something like this seems to work:
```
# recipes/_dummy/meta.yaml
package:
  name: _dummy
``` 

As far as I understood, this problem only exists when running `simulate_travis.py`. Please check pull request #3180

**Comment by daler** [comment]
Thanks, I'll check out #3180. 

What about if `repodata.json` has these contents? (and then bzip2 it as well):

```
{
  "info": {},
  "packages": {}
}
```

**Comment by jfallmann** [comment]
@daler @tetianakh 
For whatever reason, I can now run simulate_travis without errors, maybe changes in the docker image?
I'm still using the bioconda_utils hack, so unless this is solved somehow I suggest to keep the issue open

**Comment by jfallmann** [comment]
@daler @tetianakh I tried again today, and after a fresh pull from bioconda-recipes master and 
./simulate-travis.py --set-channel-order
./simulate-travis.py --install-requirements
I can now successfully run simulate-travis, still with the bioconda_utils hack from @tetianakh 
Thanks for your help!

**Comment by brunogrande** [comment]
@tetianakh: Thanks for your work on this issue!
